feat: Add support for SQS events in Amazon.Lambda.Annotations

96malhar opened the pull request:
*Issue #, if available:*
DOTNET-5526

*Description of changes:*
This PR adds support for SQS events in Amazon.Lambda.Annotations

* Added new `SQSEventAttribute` to setup event source mapping between Lambda functions and SQS queues.
* Refactored the source generator and added `LambdaFunctionValidator` class that performs all validation on the `LambdaFunctionModel` in a centralized location.
* Added new `DiagnosticDescriptors` to indicate invalid usage of `SQSEventAttribute`
* Updated `CloudFormationWriter` to parse the `SQSEventAttribute` and set up event source mapping in the `serverless.template`
* Added new unit and integration tests.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

normj commented on Libraries/test/TestServerlessApp/SQSEventExamples/InvalidSQSEvents.cs.error:
```diff
@@ -0,0 +1,41 @@
+        }
+
+        [LambdaFunction]
+        [RestApi(LambdaHttpMethod.Get, "/")]
```
Can we not have this sample. It doesn't make logical sense because the event is between SQS and API Gateway or so similar. We get users asking for a single function that can handle events from both SQS and API Gateway but there isn't really a good way to implement that. This example my give a false sense that it could work.

96malhar replied:
All functions (including this one) that are written in the `TestServerlessApp/SQSEventExamples/InvalidSQSEvents.cs.error` are invalid. This file is sent as input to the source generator unit tests and we assert that compilation errors are thrown with the appropriate diagnostic message.

In case of multiple events added to a Lambda method, we would throw a compilation error with the [MultipleEventsNotSupported](https://github.com/aws/aws-lambda-dotnet/blob/0558191fdff4235b5dbd2b2ba80633edd94c10ab/Libraries/src/Amazon.Lambda.Annotations.SourceGenerator/Diagnostics/DiagnosticDescriptors.cs#L27) diagnostic message.

Refer to the `VerifyInvalidSQSEvents_ThrowsCompilationErrors` unit test added in this PR.

normj replied:
You are right the file name is even called `InvalidSQSEvents.

normj commented on Libraries/src/Amazon.Lambda.Annotations/Amazon.Lambda.Annotations.csproj:
```diff
@@ -4,8 +4,13 @@
+
+  <PropertyGroup>
+    <Nullable>annotations</Nullable>
+    <LangVersion>10.0</LangVersion>
```
What was the reason this was changed? I get nervous about this because the package also has to run in .NET Framework which doesn't support all language version 10 features. Unless it is required I would recommend not doing this.

96malhar replied:
I wanted to set the `Filters` property in `SQSEventAttribute` to a be a nullable string (`string?`). However, nullable reference types are only supported in C# 8.0+. 

I have reverted this change and resorted to using a non-nullable string for `Filters` but with a default value of `null`

96malhar commented on Libraries/src/Amazon.Lambda.Annotations.SourceGenerator/Amazon.Lambda.Annotations.SourceGenerator.csproj:
```diff
@@ -3,6 +3,8 @@
   <PropertyGroup>
     <AssemblyVersion>1.4.0</AssemblyVersion>
     <TargetFramework>netstandard2.0</TargetFramework>
+    <AssemblyOriginatorKeyFile>..\..\..\buildtools\public.snk</AssemblyOriginatorKeyFile>
+    <SignAssembly>true</SignAssembly> 
```
The `SQSEventAttribute` defined in `Amazon.Lambda.Annotations` contains some internal methods to indicate whether the properties of this attribute were set or not. The `Amazon.Lambda.Annotation.SourceGenerator` needs to access these internal methods. Both these assemblies need to be strongly signed for the `InternalsVisibleTo` to take effect.

96malhar commented on Libraries/src/Amazon.Lambda.Annotations.SourceGenerator/Models/LambdaMethodModel.cs:
```diff
@@ -99,7 +114,7 @@ public bool ReturnsIHttpResults
         /// Gets or sets type of Lambda event
         /// </summary>
-        public List<EventType> Events { get; set; }
+        public HashSet<EventType> Events { get; set; }
```
We only care about number of unique events and therefore the type is changed from `List` to `HashShet`

96malhar commented on Libraries/src/Amazon.Lambda.Annotations.SourceGenerator/Templates/LambdaFunctionTemplate.tt:
```diff
@@ -60,7 +60,7 @@ this.Write(new FieldsAndConstructor(_model).TransformText());
         this.Write(new APIGatewayInvoke(_model, apiParameters.ParameterSignature).TransformText());
     }
-    else if (_model.LambdaMethod.Events.Count == 0)
+    else
```
Currently we only support 2 event types - `APIGatewayEvents` and `SQSEvents`. 

Since `SQSEvents` does not require any special code generation, the generated method body will be same as when the Lambda method contains no events.

96malhar commented on Libraries/src/Amazon.Lambda.Annotations/Amazon.Lambda.Annotations.csproj:
```diff
@@ -4,8 +4,10 @@
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
-  </PropertyGroup>	
-	
+    <AssemblyOriginatorKeyFile>..\..\..\buildtools\public.snk</AssemblyOriginatorKeyFile>
+    <SignAssembly>true</SignAssembly>
```
The `SQSEventAttribute` defined in `Amazon.Lambda.Annotations` contains some internal methods to indicate whether the properties of this attribute were set or not. The `Amazon.Lambda.Annotation.SourceGenerator` needs to access these internal methods. Both these assemblies need to be strongly signed for the `InternalsVisibleTo` to take effect.

96malhar commented on Libraries/src/Amazon.Lambda.Annotations/SQS/SQSEventAttribute.cs:
```diff
@@ -0,0 +1,87 @@
+    /// This attribute defines the SQS event source configuration for a Lambda function.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+    public class SQSEventAttribute : Attribute
```
Except for `Queue` all other properties are optional. 

.NET attributes cannot be nullable. To work around this, I added nullable backing fields to all optional properties and added an internal `Is<PropertyName>Set` method to identify which properties were explicitly set the customer.
These internal methods are used by the `CloudFormationWriter` while deciding which properties to write in the CF template.

ashovlin replied:
Please add this context to the code itself.

96malhar commented on Libraries/test/TestServerlessApp/SQSEventExamples/ValidSQSEvents.cs.txt:
```diff
@@ -0,0 +1,27 @@
+
+namespace TestServerlessApp.SQSEventExamples
+{
+    public class ValidSQSEvents
```
These functions represent valid usage of the `SQSEventAttribute`. However, I have added them in a `.txt` file since I do not want to deploy these function during our integration tests. This file is only in source generator unit tests. Refer to `VerifyValidSQSEvents` unit test added in this PR.

I have added a separate function that will be deployed and used for integ tests. Refer to `TestServerlessApp/SqsMessageProcessing.cs` in this PR.

96malhar commented on Libraries/src/Amazon.Lambda.Annotations/SQS/SQSEventAttribute.cs:
```diff
@@ -0,0 +1,87 @@
+        internal bool IsMaximumBatchingWindowInSecondsSet => maximumBatchingWindowInSeconds.HasValue;
+
+        /// <summary>
+        /// A collection of semicolon (;) separated strings where each string denotes a pattern. 
```
The original design had this specified as a comma separated string. However, SQS event filters are often JSON blobs. Therefore I changed this to be a semicolon separated string. 

SQS event filters - https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventfiltering.html#filtering-SQS

normj commented on Libraries/src/Amazon.Lambda.Annotations.SourceGenerator/Diagnostics/DiagnosticDescriptors.cs:
```diff
@@ -117,5 +117,26 @@ public static class DiagnosticDescriptors
+
+        public static readonly DiagnosticDescriptor ApiParametersOnNonApiFunction = new DiagnosticDescriptor(id: "AWSLambda0115",
+            title: "Invalid Usage of API Parameters",
+            messageFormat: "The lambda function parameters are annotated with HTTP API attributes but the lambda function itself is not annotated with an HTTP API attribute",
```
Capitalize the word `Lambda` in both places you use it.

normj reviewed (approved):
I just have one small comment but I don't need to review it again.

ashovlin reviewed (commented):
Could you add a section to the README covering how to use the new attribute?

ashovlin commented on Libraries/src/Amazon.Lambda.Annotations.SourceGenerator/Validation/LambdaFunctionValidator.cs:
```diff
@@ -0,0 +1,222 @@
+            }
+        }
+
+        private static void ValidateSqsEvents(LambdaFunctionModel lambdaFunctionModel, Location methodLocation, List<Diagnostic> diagnostics)
```
Could the business logic for this be in the attribute class itself? I'd rather that live "closer" to the data itself, plus if we add a ~dozen of these over time then this file will get big.

Maybe something like `private bool TryValidate(out List<Diagnostic> errors)`?

96malhar replied:
Done

96malhar commented:
> Could you add a section to the README covering how to use the new attribute?

Updated the Annotation README.

ashovlin commented:
Could code like this return a cleaner exception without the _"This is a bug."_ language? Like we do for an invalid return type or function signature?

```
    [LambdaFunction]
    [SQSEvent("queueURL")]
    public async Task<SQSBatchResponse> SQSHandler(SQSEvent evnt, ILambdaContext context)
    {
        return new SQSBatchResponse();
    }
```
```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error	AWSLambda0001	This is a bug. Please run the build with detailed verbosity (dotnet build --verbosity detailed) and file a bug at https://github.com/aws/aws-lambda-dotnet with the build output and stack trace ARN is in incorrect format. ARN format is: arn:<partition>:<service>:<region>:<account-id>:<resource>   at Amazon.Lambda.Annotations.SourceGenerator.Writers.CloudFormationWriter.GetResourceNameFromArn(String arn)
   at Amazon.Lambda.Annotations.SourceGenerator.Writers.CloudFormationWriter.ProcessSqsAttribute(ILambdaFunctionSerializable lambdaFunction, SQSEventAttribute att)
   at Amazon.Lambda.Annotations.SourceGenerator.Writers.CloudFormationWriter.ProcessLambdaFunctionEventAttributes(ILambdaFunctionSerializable lambdaFunction)
   at Amazon.Lambda.Annotations.SourceGenerator.Writers.CloudFormationWriter.ApplyReport(AnnotationReport report)
   at Amazon.Lambda.Annotations.SourceGenerator.Generator.Execute(GeneratorExecutionContext context).	AnnotationSQS	C:\Users\shovlia\source\repos\AnnotationSQS\AnnotationSQS\AnnotationSQS.csproj	1	Active	
```

ashovlin commented:
I have this definition
```
    [LambdaFunction]
    [SQSEvent("arn:aws:sqs:us-east-1:721781603755:shovlia-standard")]
    [SQSEvent("arn:aws:sqs:us-east-1:721781603755:shovlia-standard-2", BatchSize = 100, Enabled = false)]
    public async Task<SQSBatchResponse> SQSHandler(SQSEvent evnt, ILambdaContext context)
    {
        return new SQSBatchResponse();
    }
```

It builds okay locally, but then the deployment fails with

```
Error getting status of change set: Failed to create CloudFormation change set: Transform AWS::Serverless-2016-10-31 failed with: Invalid Serverless Application Specification document. Number of errors found: 1. Resource with id [AnnotationSQSFunctionsSQSHandlerGeneratedshovlia-standard] is invalid. Logical ids must be alphanumeric.

```

ashovlin commented:
I have this definition, using actual queue arns
```
    [LambdaFunction]
    [SQSEvent("arn:aws:sqs:us-east-1:<redacted>:shovlia-standard")]
    [SQSEvent("arn:aws:sqs:us-east-1:<redacted>:shovlia-standard-2", BatchSize = 100, Enabled = false)]
    public async Task<SQSBatchResponse> SQSHandler(SQSEvent evnt, ILambdaContext context)
    {
        return new SQSBatchResponse();
    }
```

It builds okay locally, but then the deployment fails with

```
Error getting status of change set: Failed to create CloudFormation change set: Transform AWS::Serverless-2016-10-31 failed with: Invalid Serverless Application Specification document. Number of errors found: 1. Resource with id [AnnotationSQSFunctionsSQSHandlerGeneratedshovlia-standard] is invalid. Logical ids must be alphanumeric.

```

ashovlin commented on Libraries/src/Amazon.Lambda.Annotations/README.md:
```diff
@@ -709,6 +722,98 @@ public class Functions_Resize_Generated
+[SQSEvent("@TestQueue", BatchSize = 50, MaximumConcurrency = 5, MaximumBatchingWindowInSeconds = 5, Filters = "{ \"body\" : { \"RequestCode\" : [ \"BBBB\" ] } }")]
+public SQSBatchResponse HandleMessage(SQSEvent evnt, ILambdaContext lambdaContext)
+{
+	lambdaContext.Logger.Log($"Received {evnt.Records.Count} messages");
```
nit: can you fix the indentation here?

96malhar replied:
Done

96malhar commented:
> Could code like this return a cleaner exception without the _"This is a bug."_ language? Like we do for an invalid return type or function signature?
> 
> ```
>     [LambdaFunction]
>     [SQSEvent("queueURL")]
>     public async Task<SQSBatchResponse> SQSHandler(SQSEvent evnt, ILambdaContext context)
>     {
>         return new SQSBatchResponse();
>     }
> ```
> 
> ```
> Severity	Code	Description	Project	File	Line	Suppression State	Details
> Error	AWSLambda0001	This is a bug. Please run the build with detailed verbosity (dotnet build --verbosity detailed) and file a bug at https://github.com/aws/aws-lambda-dotnet with the build output and stack trace ARN is in incorrect format. ARN format is: arn:<partition>:<service>:<region>:<account-id>:<resource>   at Amazon.Lambda.Annotations.SourceGenerator.Writers.CloudFormationWriter.GetResourceNameFromArn(String arn)
>    at Amazon.Lambda.Annotations.SourceGenerator.Writers.CloudFormationWriter.ProcessSqsAttribute(ILambdaFunctionSerializable lambdaFunction, SQSEventAttribute att)
>    at Amazon.Lambda.Annotations.SourceGenerator.Writers.CloudFormationWriter.ProcessLambdaFunctionEventAttributes(ILambdaFunctionSerializable lambdaFunction)
>    at Amazon.Lambda.Annotations.SourceGenerator.Writers.CloudFormationWriter.ApplyReport(AnnotationReport report)
>    at Amazon.Lambda.Annotations.SourceGenerator.Generator.Execute(GeneratorExecutionContext context).	AnnotationSQS	C:\Users\shovlia\source\repos\AnnotationSQS\AnnotationSQS\AnnotationSQS.csproj	1	Active	
> ```

Added a check to verify that the `Queue` property has a valid ARN structure and returns a clear diagnostic error when the validation fails.

96malhar commented:
> I have this definition, using actual queue arns
> 
> ```
>     [LambdaFunction]
>     [SQSEvent("arn:aws:sqs:us-east-1:<redacted>:shovlia-standard")]
>     [SQSEvent("arn:aws:sqs:us-east-1:<redacted>:shovlia-standard-2", BatchSize = 100, Enabled = false)]
>     public async Task<SQSBatchResponse> SQSHandler(SQSEvent evnt, ILambdaContext context)
>     {
>         return new SQSBatchResponse();
>     }
> ```
> 
> It builds okay locally, but then the deployment fails with
> 
> ```
> Error getting status of change set: Failed to create CloudFormation change set: Transform AWS::Serverless-2016-10-31 failed with: Invalid Serverless Application Specification document. Number of errors found: 1. Resource with id [AnnotationSQSFunctionsSQSHandlerGeneratedshovlia-standard] is invalid. Logical ids must be alphanumeric.
> ```

I have added a new `ResourceName` property in the `SQSEvent` attribute. It sets the CloudFormation resource name for the SQS event source mapping. By default this is set to the SQS queue name if `Queue` is set to an SQS queue ARN. It will sanitize the SQS queue name by stripping out non-alphanumeric characters. If `Queue` is set to an existing CloudFormation resource, than that is used as the default value without the "@" prefix.

Also added new unit tests that exercises this logic.

ashovlin commented:
Could we throw a validation error in advance if invalid settings are applied to a fifo queue?

```
6/12/2024 11:15:45 AM	AWS::Lambda::EventSourceMapping	MySQSLambdaFunctionshovliafifo			Resource handler returned message: "Invalid request provided: Batching window is not supported for FIFO queues (Service: Lambda, Status Code: 400, Request ID: 5df30f34-e266-44c9-b4a1-061da42df499)" (RequestToken: 539661a7-ca6f-b190-ad75-5bd099921c77, HandlerErrorCode: InvalidRequest)
```

```
[SQSEvent("arn:aws:sqs:us-east-1:<redacted>:shovlia.fifo", BatchSize = 100, MaximumBatchingWindowInSeconds = 5)]
public void SQSHandler(SQSEvent evnt, ILambdaContext context)
```

96malhar commented on Libraries/src/Amazon.Lambda.Annotations/SQS/SQSEventAttribute.cs:
```diff
@@ -0,0 +1,153 @@
+        {
+            var validationErrors = new List<string>();
+
+            if (IsBatchSizeSet && (BatchSize < 1 || BatchSize > 10000))
+            {
+                validationErrors.Add($"{nameof(SQSEventAttribute.BatchSize)} = {BatchSize}. It must be between 1 and 10000");
+            }
```
Added the validation to check that `MaximumBatchingWindowInSeconds` is at least 1 when `BatchSize` > 10. Also added new tests asserting this behavior.

96malhar commented:
> Could we throw a validation error in advance if invalid settings are applied to a fifo queue?
> 
> ```
> 6/12/2024 11:15:45 AM	AWS::Lambda::EventSourceMapping	MySQSLambdaFunctionshovliafifo			Resource handler returned message: "Invalid request provided: Batching window is not supported for FIFO queues (Service: Lambda, Status Code: 400, Request ID: 5df30f34-e266-44c9-b4a1-061da42df499)" (RequestToken: 539661a7-ca6f-b190-ad75-5bd099921c77, HandlerErrorCode: InvalidRequest)
> ```
> 
> ```
> [SQSEvent("arn:aws:sqs:us-east-1:<redacted>:shovlia.fifo", BatchSize = 100, MaximumBatchingWindowInSeconds = 5)]
> public void SQSHandler(SQSEvent evnt, ILambdaContext context)
> ```

Added the necessary checks for FIFO specific validations. Discussed offline and concluded that we will only perform these checks when the `Queue` property in the `SQSEvent` attribute points to an ARN that ends in `.fifo`.

ashovlin commented on Libraries/src/Amazon.Lambda.Annotations/SQS/SQSEventAttribute.cs:
```diff
@@ -134,6 +136,26 @@ internal List<string> Validate()
+            if (IsBatchSizeSet && BatchSize > 10 && (!IsMaximumBatchingWindowInSecondsSet || MaximumBatchingWindowInSeconds < 1))
+            {
+                validationErrors.Add($"{nameof(SQSEventAttribute.MaximumBatchingWindowInSeconds)} is not set or set to a value less than 1. " +
+                    $"It must be set to atleast 1 when {nameof(SQSEventAttribute.BatchSize)} is greater than 10");
```
nit: missing space

96malhar replied:
Fixed
